crates/worker/src/partition/snapshot_producer.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2023 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::SystemTime;
+
+use anyhow::{anyhow, bail};
+use parking_lot::Mutex;
+use tokio::sync::oneshot::Sender;
+use tracing::{info, warn};
+
+use restate_core::TaskKind;
+use restate_partition_store::snapshots::{PartitionSnapshotMetadata, SnapshotFormatVersion};
+use restate_partition_store::PartitionStore;
+use restate_types::config::Configuration;
+use restate_types::identifiers::{PartitionId, SnapshotId};
+use restate_types::live::LiveLoad;
+use restate_types::time::MillisSinceEpoch;
+
+pub struct SnapshotProducer {
+    pub partition_id: PartitionId,
+    pub config: Box<dyn LiveLoad<Configuration> + Send>,
+    latest_snapshot: Arc<Mutex<Option<SnapshotId>>>,
+}
+
+impl SnapshotProducer {
+    pub fn new(partition_id: PartitionId, config: Box<dyn LiveLoad<Configuration> + Send>) -> Self {
+        Self {
+            partition_id,
+            config,
+            latest_snapshot: Arc::new(Mutex::new(None)),
+        }
+    }
+
+    /// Creates a snapshot in a background task, optionally sending a completion notification.
+    pub(crate) fn create_snapshot(
+        &mut self,
+        partition_store: &mut PartitionStore,
+        response_tx: Option<Sender<anyhow::Result<SnapshotId>>>,
```

Comment:
nit: Perhaps `SnapshotProducer` doesn't need to own the live configuration and we can take `&Configuration` as input here. In partition processor you can use `&Configuration::pinned()` to pass down a fresh configuration reference when calling `create_snapshot()`